### ISSUE TITLE
utils: Improve how we get a free port (without brute force)

### ISSUE THREAD

**Issue by beraldoleal** [avocado-framework/avocado#5273]
**Is your feature request related to a problem? Please describe.**
Currently, `utils.find_free_port()` is brute forcing for a free port. According to @huth about QEMU tests: _"..this takes > 70 seconds for each and every tests from tests/avocado/boot_linux.py to get canceled."_

**Describe the solution you'd like**
Implement an improved version of this method.

**Describe alternatives you've considered**
Maybe something like:

```python
with socket() as s: 
    s.bind(('',0)) 
    return s.getsockname()[1]
```  

**Additional information**
https://www.mail-archive.com/qemu-devel@nongnu.org/msg871809.html

**Comment by beraldoleal** [comment]
@huth, if you could describe what in your system is blocking this method, would be great. So far, I'm not able to reproduce this.

**Comment by huth** [comment]
I thought it might be related to the firewall or SELinux, but disabling those didn't cure the issue. So I've had a closer look now, and it seems to me like that is_port_free() function in avocado/utils/network/ports.py is completely flawed. Seems like it is always trying all protocols and families instead of returning on the first working combination - and AF_INET6 is not configured on my system, so it is finally running into an error.

So please try to disable all inet6 addresses on your system, I think you should then be able to reproduce this issue.

I can make it work here with the following patch:

```
diff --git a/avocado/utils/network/ports.py b/avocado/utils/network/ports.py
--- a/avocado/utils/network/ports.py
+++ b/avocado/utils/network/ports.py
@@ -60,7 +60,7 @@ def is_port_free(port, address):
                     if localhost:
                         return False
                 sock.close()
-        return True
+                return True
     finally:
         if sock is not None:
             sock.close()
```

... not sure whether that's a real good solution here, though ... my Python-foo is really poor, but aren't there some more issues in this function? Is the socket really closed in all cases (i.e. also the spots where the code "returns False")?

**Comment by huth** [comment]
By the way, I wonder what happens if you only have IPv6 addresses on your system, but no IPv4 addresses at all? ... I guess it's also completely failing in that case...

**Comment by berrange** [comment]
As mentioned on the QEMU list, this whole problem space would be dramatically simplified if the code for finding free ports was entirely deleted, and the callers changed to rely on the kernel auto-allocating a free port. The issue with IPv6 would simply go away, as would the inherent race conditions in this code when running  tests in parallel.

**Comment by huth** [comment]
Another issue: the code in is_port_free() seems to assume that UDP and TCP share the same port number space. That's complete non-sense, if a TCP port is already in use, you can still open the UDP port with the same number, and vice versa. I think all the related code (especially the unit test in selftests/unit/test_utils_network.py ) need a major rewrite ... :-(

**Comment by beraldoleal** [comment]
Thanks @huth for the detailed description. I understand the problem now and indeed deserves a rewrite.
